rosidl_generator_c/resource/idl__description.c.em

```diff
@@ -0,0 +1,140 @@
+
+type_description_msg = type_description_info['type_description_msg']
+all_type_descriptions = [type_description_msg['type_description']] + type_description_msg['referenced_type_descriptions']
+
+include_parts = [package_name] + list(interface_path.parents[0].parts) + [
+    'detail', convert_camel_case_to_lower_case_underscore(interface_path.stem)]
+include_base = '/'.join(include_parts)
+
+full_type_descriptions = [type_description_msg]
+for service in content.get_elements_of_type(Service):
+  full_type_descriptions.append(extract_subinterface(type_description_msg, 'request_message'))
+  full_type_descriptions.append(extract_subinterface(type_description_msg, 'response_message'))
+  full_type_descriptions.append(extract_subinterface(type_description_msg, 'event_message'))
+for action in content.get_elements_of_type(Action):
+  full_type_descriptions.append(extract_subinterface(type_description_msg, 'goal'))
+  full_type_descriptions.append(extract_subinterface(type_description_msg, 'result'))
+  full_type_descriptions.append(extract_subinterface(type_description_msg, 'feedback'))
+
+  send_goal_service = extract_subinterface(type_description_msg, 'send_goal_service')
+  full_type_descriptions.append(send_goal_service)
+  full_type_descriptions.append(extract_subinterface(send_goal_service, 'request_message'))
+  full_type_descriptions.append(extract_subinterface(send_goal_service, 'response_message'))
+  full_type_descriptions.append(extract_subinterface(send_goal_service, 'event_message'))
+
+  get_result_service = extract_subinterface(type_description_msg, 'get_result_service')
+  full_type_descriptions.append(get_result_service)
+  full_type_descriptions.append(extract_subinterface(get_result_service, 'request_message'))
+  full_type_descriptions.append(extract_subinterface(get_result_service, 'response_message'))
+  full_type_descriptions.append(extract_subinterface(get_result_service, 'event_message'))
+
+  full_type_descriptions.append(extract_subinterface(type_description_msg, 'feedback_message'))
+
+def typename_to_c(typename):
+  return typename.replace('/', '__')
+
+def static_seq(varname, values):
+  """Statically define a runtime Sequence or String type."""
+  if values:
+    return f'{{{varname}, {len(values)}, {len(values)}}}'
+  return '{NULL, 0, 0}'
+}@
+
+#include "@(include_base)__struct.h"
+
+// Declare and define all type names, field names, and default values
+@[for itype_description in all_type_descriptions]@
+@{
+td_c_typename = typename_to_c(itype_description['type_name'])
+}@
+static char @(td_c_typename)__TYPE_NAME[] = "@(itype_description['type_name'])";
+@[  for field in itype_description['fields']]@
+static char @(td_c_typename)__FIELD_NAME__@(field['name'])[] = "@(field['name'])";
+@[    if field['default_value']]@
+static char @(td_c_typename)__DEFAULT_VALUE__@(field['name'])[] = "@(escape_string(field['default_value']))";
+@[    end if]@
+@[  end for]@
+
+@[end for]@
```

Comment:
Do we really have to expand all types and type names recursively here?  That leads to a lot of duplicated strings, and is going to bloat the sizes of our binary objects quite a lot.  Can we instead `#include` the headers we need?

Reply:
This is actually a pretty big problem - we had a discussion about this in the weekly meeting. We don't know when this was built, we don't know when any dependencies of it were built.

Here's my attempt at breaking down this problem:
* We want to bake descriptions and type hashes into the code
* Type hashes depend on all recursive dependencies
* You have interface package `a_msgs`, and `b_msgs` which depends on it
* `b_msgs/B`  has a field of type `a_msgs/A`
* You build package `b_msgs`
* You change `a_msgs/A` and build `a_msgs` only
* There is now no possible way for `b_msgs/B` to have a correct baked-in type hash at code gen _or_ compilation time, the hash it has is the hash corresponding to its understanding of the full recursive type at the time it was last built
* The hash that is baked in corresponds to the description that we are also baking in - if we use a reference via headers to `a_msgs/msg/a.h` to get portions of the description - then the description that you receive will not hash out equal to the baked in type hash
* If, alternatively, we copy the values at compilation time for `std_msgs/B`, then it doesn't  matter whether we defined them here or elsewhere, the binary bloat will be equivalent.

@wjwwood I could be missing something but I believe this is the extent of the issue we discussed, and is why we said we have to bake in the full recursive type for _each_ type that we generate descriptions for.

Either that or we have to guarantee no static availability of type hashes, and rewrite to a runtime-only hasher